### DISCUSSION TITLE
Julia syntax: calling a function with kwargs

### DISCUSSION THREAD

**Discussion by JohnAAbbott** [oscar-system/Oscar.jl#2615] [General]
What is the syntax recommendation for calling a function with kwargs?
I was somewhat surprised to find the following example on the julialang website
```
julia> foo(;kwargs...) = dump(kwargs)
foo (generic function with 1 method)

julia> foo(a=1, b=2, c=3)
Base.Pairs{Symbol, Int64, Tuple{Symbol, Symbol, Symbol}, NamedTuple{(:a, :b, :c), Tuple{Int64, Int64, Int64}}}
  data: NamedTuple{(:a, :b, :c), Tuple{Int64, Int64, Int64}}
    a: Int64 1
    b: Int64 2
    c: Int64 3
  itr: Tuple{Symbol, Symbol, Symbol}
    1: Symbol a
    2: Symbol b
    3: Symbol c
```
Why is the call not `foo(;a=1, b=2, c=3)`?  It seems that putting in the semicolon produces the same result as without.
Is there a recommendation for which syntax to use in Oscar code?
[This question was triggered by line 585 in `mpoly-affine-algebras.jl` where a kwarg is specified without a preceding semicolon]

**Comment by thofma** [comment]
It is usually a good idea to separate the keyword arguments with a semicolon, since otherwise one might not always get what one wants:
```
julia> sum(i for i in 1:2, init = 5)
3

julia> sum(i for i in 1:2; init = 5)
8
```

**Reply by JohnAAbbott** [reply]
Ouch!  That's a subtle difference.  I wonder what the first call does...?  I tried `@which` but that was not so helpful.
Anyway I'll correct line 585 in `mpoly-affine-algebras.jl` (and maybe think of a regexp to find other instances where a comma was used instead of a semicolon).

**Comment by JohnAAbbott** [comment]
In light of the comment by @thofma  it seems to be a _really good idea_ to ensure that we use a semicolon before specifying kwargs, otherwise we will have code which "looks correct" but might not do what one reasonably expects (_i.e._ somewhat **evil bugs**).
A fairly simple regexp for finding source code lines which are suspicious is `([^);]*=`  (at least that seems to work in Emacs), but this probably does not work in all cases... Any suggestions for a better regexp?

**Reply by JohnAAbbott** [reply]
Here is another (Emacs) regexp: it has more false positives, but fewer false negatives.  The regexp `([^;^J]*[_a-zA-Z]+ *=`
It looks for an open bracket not followed by a semicolon, but with something looking like `identifier = ...`

**Comment by JohnAAbbott** [comment]
In PR #2614  @lgoettgens  points out that when using a semicolon, one can use short-cut assignments for kwargs (if kwarg-name and assigned value are written identically).  This makes the source code shorter... not sure if it makes the code clearer though..  Advice???
_For example:_
```
return ideal(homogenization(gens(groebner_basis(I; ordering=ordering)), var; pos=pos))
```
could be more succinctly written as
```
return ideal(homogenization(gens(groebner_basis(I; ordering)), var; pos))
```
[but then the kwargs do not stand out to the "casual reader"]

**Reply by thofma** [reply]
I think this is subjective. I personally don't find it clearer at all.

**Comment by JohnAAbbott** [comment]
Since no-one has expressed a strong opinion either way, I'll go with what I prefer (in my code), namely `pos=pos` _etc._
If we later decide that the shorter form is better, it should be easy to automatically convert `pos=pos` into just `pos`.
